library/sonar_plugin.py

```diff
@@ -0,0 +1,429 @@
+        else:
+            output = 'Lower'
+
+    return output
+
+
+def get_link_from_repo(repo_url, key, version):
+    final_url = 'not-found'
+    html_doc = urllib2.urlopen(repo_url).read()
+    parser = LinksParser()
+    parser.feed(html_doc)
+    all_data = parser.data
+
+    for data in all_data:
+        match = re.search('sonar-(.*)-plugin/', data)
+        if match is not None:
+            plugin = match.group(1)
+            current_key = re.sub('-', '', plugin)
+            if current_key == key:
+                ready_data = re.sub(r'.*sonar', 'sonar', data)
+                final_url = repo_url + data.lstrip() + re.sub('/', '', ready_data) + '-' + version + '.jar'
+                return final_url
+
+    return final_url
+
+
+def is_plugin_update_available(module, sonar_url, updates_list_api, installed_list_api):
+    is_available = False
+    version_installed = 'not-found'
+    url = sonar_url + installed_list_api
+    response = requests.get(url, auth=HTTPBasicAuth(module.params['username'],
+                                                    module.params['password']))
+    json_obj = response.json()
+
+    for plugin in json_obj['plugins']:
+        if plugin['name'] == module.params['name']:
+            version_installed = plugin['version']
+
+    comparison = compare_plugins_version(module.params['version'],
+                                         version_installed)
+
+    if comparison == 'Equal':
+        module.exit_json(changed=False, stdout='FAILED: Version already installed')
+    elif comparison == 'Lower':
+        is_available = True
+    else:
+        url = sonar_url + updates_list_api
+        response = requests.get(url, auth=HTTPBasicAuth(module.params['username'], module.params['password']))
+        json_obj = response.json()
+        for plugin in json_obj['plugins']:
+            if plugin['name'] == module.params['name']:
+                for update in plugin['updates']:
+                    if compare_plugins_version(module.params['version'], update['release']['version']) == 'Equal':
+                        if not update['requires'] and update['status'] == 'COMPATIBLE':
+                            is_available = True
+
+    return is_available
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            name=dict(type='str'),
+            state=dict(type='str', choices=['installed', 'removed', 'latest']),
+            version=dict(type='str'),
+            custom_url=dict(type='str'),
+            hostname=dict(default='localhost', type='str'),
+            username=dict(default='admin', type='str'),
+            password=dict(default='admin', type='str', no_log=True),
+            pending_dir=dict(default='/usr/local/sonar/extensions/downloads/', type='str'),
+            sonar_port=dict(default='9000', type='int')
+        ),
+        mutually_exclusive=[['custom_url', 'state']]
+    )
+
+    available_list_api = '/api/plugins/available'
```

Comment:
If you only use the string once\few times, I think it safe use it as is instead of putting it in a variable.

Reply:
done